File: solver.cpp
#include <iostream>
#include <vector>

using namespace std;

const int SIZE = 6;
const int BOX_ROWS = 2;
const int BOX_COLS = 3;


void printGrid(const vector<vector<int>>& board) {
    for (int i = 0; i < SIZE; i++) {
        if (i % BOX_ROWS == 0 && i != 0) {
            cout << "-----------------" << endl;
        }
        for (int j = 0; j < SIZE; j++) {
            if (j % BOX_COLS == 0 && j != 0) {
                cout << "| ";
            }
            cout << board[i][j] << " ";
        }
        cout << endl;
    }
}


bool isValid(const vector<vector<int>>& board, int row, int col, int num) {
    
    for (int x = 0; x < SIZE; x++) {
        if (board[row][x] == num) return false;
    }

    
    for (int x = 0; x < SIZE; x++) {
        if (board[x][col] == num) return false;
    }

    
    int startRow = row - row % BOX_ROWS;
    int startCol = col - col % BOX_COLS;
    
    for (int i = 0; i < BOX_ROWS; i++) {
        for (int j = 0; j < BOX_COLS; j++) {
            if (board[i + startRow][j + startCol] == num) return false;
        }
    }

    return true;
}


bool solveSudoku(vector<vector<int>>& board) {
    for (int row = 0; row < SIZE; row++) {
        for (int col = 0; col < SIZE; col++) {
            if (board[row][col] == 0) {
                for (int num = 1; num <= SIZE; num++) {
                    if (isValid(board, row, col, num)) {
                        board[row][col] = num;

                        if (solveSudoku(board)) return true;

                        board[row][col] = 0; 
                    }
                }
                return false;
            }
        }
    }
    return true;
}

int main() {
    
    vector<vector<int>> grid(SIZE, vector<int>(SIZE, 0));

    cout << "======================================================" << endl;
    cout << "             Custom 6x6 Sudoku Solver                 " << endl;
    cout << "======================================================" << endl;
    cout << "Enter your 6x6 puzzle row by row." << endl;
    cout << "Use numbers 1-6 for filled cells, and 0 for empty spaces." << endl;
    cout << "Separate numbers in a row with spaces.\n" << endl;

    
    for (int i = 0; i < SIZE; i++) {
        cout << "Enter Row " << i + 1 << " (6 numbers): ";
        for (int j = 0; j < SIZE; j++) {
            cin >> grid[i][j];
        }
    }

    cout << "\n-----------------------------" << endl;
    cout << "Your Inputted Puzzle:" << endl;
    printGrid(grid);
    cout << "-----------------------------\n" << endl;

    cout << "Solving custom puzzle...\n" << endl;

    if (solveSudoku(grid)) {
        cout << "=============================" << endl;
        cout << "      SOLVED PUZZLE" << endl;
        cout << "=============================" << endl;
        printGrid(grid);
    } else {
        cout << "No valid solution exists for this puzzle layout." << endl;
    }

    return 0;
}

File: Solver.py
# 0 represents empty cells that the solver needs to fill
grid = [
    [4, 0, 0, 0, 0, 1],
    [0, 0, 3, 0, 6, 0],
    [0, 2, 0, 0, 0, 5],
    [1, 0, 0, 0, 3, 0],
    [0, 4, 0, 2, 0, 0],
    [5, 0, 0, 0, 0, 6]
]

def print_grid(board):
    """Prints the 6x6 board with clear visual boundaries."""
    for i in range(6):
        # Print a horizontal line after every 2 rows
        if i % 2 == 0 and i != 0:
            print("-" * 17)
        for j in range(6):
            # Print a vertical line after every 3 columns
            if j % 3 == 0 and j != 0:
                print("| ", end="")
            print(f"{board[i][j]} ", end="")
        print()

def is_valid(board, row, col, num):
    """Checks if placing 'num' at board[row][col] is valid."""
    # 1. Check the row for duplicates
    for x in range(6):
        if board[row][x] == num:
            return False

    # 2. Check the column for duplicates
    for x in range(6):
        if board[x][col] == num:
            return False

    # 3. Check the 2x3 box
    # Find the starting row (0, 2, or 4) and column (0 or 3) index of the 2x3 box
    start_row = row - row % 2
    start_col = col - col % 3
    for i in range(2):
        for j in range(3):
            if board[i + start_row][j + start_col] == num:
                return False

    return True

def solve_sudoku(board):
    """Solves the Sudoku board using backtracking recursion."""
    for row in range(6):
        for col in range(6):
            # Look for an empty cell
            if board[row][col] == 0:
                # Try placing numbers 1 through 6
                for num in range(1, 7):
                    if is_valid(board, row, col, num):
                        board[row][col] = num  # Tentatively place number

                        # Recursively try to solve the rest of the board
                        if solve_sudoku(board):
                            return True

                        # If it fails later, UNDO the choice (Backtrack)
                        board[row][col] = 0
                
                return False  # Trigger backtracking to previous steps
    return True  # Board completely solved

if __name__ == "__main__":
    print("Original 6x6 Sudoku Puzzle:")
    print_grid(grid)
    print("\nSolving...\n")
    
    if solve_sudoku(grid):
        print("Solved 6x6 Sudoku Puzzle:")
        print_grid(grid)
    else:
        print("No solution exists for this puzzle.")